### DISCUSSION TITLE
Undefined reference to abort

### DISCUSSION THREAD

**Discussion by ssnover** [Rahix/avr-hal#334] [Q&A] [answered]
I've been trying to get a minimal example building and have hit a wall trying to get the final project to link together.

[Full project repo is here](https://github.com/ssnover/arduino-blink)

But I'm compiling with recently nightly, have pulled in `avr-std-stub` in Cargo.toml and in main.rs have `extern crate avr_std_stub;`. Finally, I've specified in my `.cargo/config.toml` to build libraries and set my target:
```toml
[build]
target = "avr-unknown-gnu-atmega328"

[unstable]
build-std = ["core", "compiler_builtins"]

```
However, despite all this, I'm getting errors on build with `cargo +nightly build --release`:

```
ssnover:arduino-blink$ cargo +nightly build --release
   Compiling arduino-blink v0.1.0 (/home/ssnover/dev/arduino-blink)
WARN rustc_codegen_ssa::back::link Linker does not support -no-pie command line option. Retrying without.
error: linking with `avr-gcc` failed: exit status: 1
  |
  = note: "avr-gcc" "-mmcu=atmega328" "/home/ssnover/dev/arduino-blink/target/avr-unknown-gnu-atmega328/release/deps/arduino_blink-7b03611be97ed54b.arduino_blink.7cb9b83c-cgu.0.rcgu.o" "-Wl,--as-needed" "-L" "/home/ssnover/dev/arduino-blink/target/avr-unknown-gnu-atmega328/release/deps" "-L" "/home/ssnover/dev/arduino-blink/target/release/deps" "-L" "/home/ssnover/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/avr-unknown-gnu-atmega328/lib" "-Wl,--start-group" "-Wl,--end-group" "-Wl,-Bstatic" "/home/ssnover/dev/arduino-blink/target/avr-unknown-gnu-atmega328/release/deps/libcompiler_builtins-4be1e100b4795ac1.rlib" "-Wl,-Bdynamic" "-lgcc" "-Wl,-znoexecstack" "-L" "/home/ssnover/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/avr-unknown-gnu-atmega328/lib" "-o" "/home/ssnover/dev/arduino-blink/target/avr-unknown-gnu-atmega328/release/deps/arduino_blink-7b03611be97ed54b.elf" "-Wl,--gc-sections" "-Wl,-O1" "-nodefaultlibs"
  = note: /home/ssnover/dev/arduino-blink/target/avr-unknown-gnu-atmega328/release/deps/arduino_blink-7b03611be97ed54b.arduino_blink.7cb9b83c-cgu.0.rcgu.o: In function `main':
          arduino_blink.7cb9b83c-cgu.0:(.text.main+0x4): undefined reference to `abort'
          /home/ssnover/dev/arduino-blink/target/avr-unknown-gnu-atmega328/release/deps/arduino_blink-7b03611be97ed54b.arduino_blink.7cb9b83c-cgu.0.rcgu.o: In function `arduino_blink::__avr_device_rt_main::hcc00f93d7920e3e3':
          arduino_blink.7cb9b83c-cgu.0:(.text._ZN13arduino_blink20__avr_device_rt_main17hcc00f93d7920e3e3E+0x9c): undefined reference to `abort'
          /home/ssnover/dev/arduino-blink/target/avr-unknown-gnu-atmega328/release/deps/arduino_blink-7b03611be97ed54b.arduino_blink.7cb9b83c-cgu.0.rcgu.o: In function `core::panicking::panic::h3bd82c06970c382b':
          arduino_blink.7cb9b83c-cgu.0:(.text.unlikely._ZN4core9panicking5panic17h3bd82c06970c382bE+0x4): undefined reference to `abort'
          collect2: error: ld returned 1 exit status
          
  = help: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
  = note: use the `-l` flag to specify native libraries to link
  = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)

error: could not compile `arduino-blink` due to previous error
```

Is `abort` something that needs to be defined in my project since it's not being provided by `libcore` or is this supposed to be similar to `eh_personality` where there's a crate that's necessary to make the ecosystem work?

**Comment by Rahix** [comment, marked as answer]
Hi, what exactly are you trying to do?  Are you just interested in getting Rust on AVR running?  Then please check out the guide in the README of this project: [Quickstart](https://github.com/Rahix/avr-hal#quickstart).  It links you to a project template that has everything set up correctly already.

If you are specifically trying to get `avr-std-stub` working, maybe you need to tell me more about the reasons...

---

That out of the way, a few technical comments about your question.  You can ignore them if you are just starting out...

- `avr-std-stub` is not needed on new rustc versions.  The panic handler can either be pulled in from one of the existing implementations like [`panic-halt`](https://crates.io/crates/panic-halt) or any of the other `panic-*` crates.  Or you can define it yourself with custom behavior as seen in this `avr-hal` example: [`uno-panic.rs`](https://github.com/Rahix/avr-hal/blob/main/examples/arduino-uno/src/bin/uno-panic.rs)
- `eh_personality` is about unwinding.  On AVR you probably shouldn't do unwinding panics.  You can switch to aborting panics by setting `panic = "abort"` in your cargo profile.  After doing that, the lang item `eh_personality` is also not needed anymore.
- `abort()` is usually defined by the runtime.  In the case of AVR, that is `avr-libc`.  Right now, we are still depending on `avr-libc` for such runtime features but this will eventually change (see https://github.com/Rahix/avr-device/issues/76)
- Finally, I see you're explicitly mentioning `compiler-builtins` in your `build-std` config.  I've never seen that being required, was there a reason why you added it?

**Reply by ssnover** [reply]
Admittedly some of the changes I made to how the crate was building was trial and error based on error messages I was getting and experience I've had previously with getting msp430 and cortex-m projects to build in Rust. This version was the farthest I got, but some of it may be is now hindering the correct solution.

I'll scan the template in the quickstart for differences and get back to you. I was using the avr-rust book to get a basic project started and I think a fair bit of it is currently out-of-date based on my experimentation and some of your technical comments above.

**Reply by ssnover** [reply]
Relative to what I had before I've now swapped `avr-std-stub` for `panic-halt`, added `panic = "abort"` to my build profile, and removed `compiler-builtins`. I'm still getting a linking error on `abort` which implies to me that `avr-libc`'s not being found during linking.

I've got the correct archive file at `/usr/lib/avr/lib/avrxmega3/libc.a`. But don't see it in my linker invocation to `avr-gcc` that's being printed (just `-lgcc` so maybe that's supposed to be linking libc?).

**Reply by ssnover** [reply]
Alright, the final unlock that I admittedly don't quite understand is pulling the target spec file from the `avr-hal-template` repo. It seems like the built-in target `avr-unknown-gnu-atmega328` doesn't correctly link `avr-libc` which is an unfortunate footgun. Meanwhile the one called out in the spec `avr-unknown-unknown` I guess properly finds it.

**Reply by ssnover** [reply]
My compiler version was `cargo 1.65.0-nightly (4ed54cecc 2022-08-27)` for anyone who might be following along later.
